src/arch/metal.rs

```diff
@@ -531,6 +533,131 @@ impl Architecture for MetalArch {
+        // Cannot use a raw_trap_handler because of MPRV=1, addresses would be translated
+        let trap: usize = _mprv_trap_handler as usize;
+
+        // Zero out mcause to check if a trap occured during emulation
+        Self::write_csr(Csr::Mcause, 0);
+
+        let saved_satp = Self::read_csr(Csr::Satp);
+        let saved_mpp = parse_mpp_return_mode(Self::read_csr(Csr::Mstatus));
+
+        Self::set_mpp(parse_mpp_return_mode(ctx.csr.mstatus));
+        Self::write_csr(Csr::Satp, ctx.csr.satp);
+        Self::sfence_vma(None, None);
+
+        let mut _rd_value: usize; //without "_" compiler warns that the value is never read in case of a store macro
+        let addr: usize;
+        let mut cause: usize;
+        let mut mtval: usize;
+        let mut mstatus: usize;
+        let mut mip: usize;
+        let fw_pc = ctx.trap_info.mepc;
+
+        macro_rules! construct_asm {
+            ($instr:literal) => {
+                asm!(
+                    "csrr {1}, mtvec",
+                    "csrw mtvec, {mtvec}",
+
+                    "li {0}, 1",           // Set MPRV bit to 1
+                    "slli {0}, {0}, 17",
+                    "csrs mstatus, {0}",
+
+                    concat!($instr, " {rd}, 0({addr})"),
+
+                    "csrc mstatus, {0}",
+                    "csrw mtvec, {1}",
+                    out(reg) _,
+                    out(reg) _,
+                    out("t1") cause,
+                    out("t2") mtval,
+                    out("t4") mstatus,
+                    out("t6") mip,
+                    mtvec = in(reg) trap,
+                    addr = in(reg) addr,
+                    rd = inout(reg) _rd_value,
+                )
+            };
+        }
+
+        match instr {
+            Instr::Load {
+                rd,
+                rs1,
+                imm,
+                len,
+                is_compressed,
+                is_unsigned,
+            } => {
+                addr = utils::calculate_addr(ctx.get(rs1), imm);
+                _rd_value = 0;
+
+                match (len, is_unsigned) {
+                    (Width::Byte, false) => construct_asm!("lb"),
+                    (Width::Byte2, false) => construct_asm!("lh"),
+                    (Width::Byte4, false) => construct_asm!("lw"),
+                    (Width::Byte8, false) => construct_asm!("ld"),
+                    (Width::Byte, true) => construct_asm!("lbu"),
+                    (Width::Byte2, true) => construct_asm!("lhu"),
+                    (Width::Byte4, true) => construct_asm!("lwu"),
+                    _ => panic!("Unknown load instruction"),
+                };
+
+                if Self::read_csr(Csr::Mcause) != 0 {
+                    // Unsure if a trap might happen here with a correct emulation, probably not?
+
+                    ctx.trap_info.mcause = cause;
+                    ctx.trap_info.mstatus = mstatus;
+                    ctx.trap_info.mtval = mtval;
```

Comment:
That would be virtual address. In theory, the only way it could get anything other that this would be catching an `Illegal instruction` exception, which should be impossible.